blacklists.py

```diff
@@ -140,17 +147,174 @@ def exists(self, item: Union[str, dict]):
                 splat = x.split('\t')
                 if len(splat) == 3 and splat[2].strip() == item:
                     return True, i
+        return False, -1
+
+
+class YAMLParserCIDR(BlacklistParser):
+    """
+    YAML parser for IP blacklist (name suggests we should move to proper CIDR eventually).
+
+    Base class for parsers for YAML files with simple schema validation.
+    """
+    # Remember to update the schema version if any of this needs to be changed
+    SCHEMA_VERSION = '2019120601'  # yyyy mm dd id
+    SCHEMA_VARIANT = 'yaml_cidr'
+    SCHEMA_PRIKEY = 'ip'
+
+    def __init__(self, filename):
+        super().__init__(filename)
+
+    def _parse(self, keep_disabled=False):
+        with open(self._filename, 'r', encoding='utf-8') as f:
+            y = yaml.safe_load(f)
+        assert y['Schema'] == self.__class__.SCHEMA_VARIANT
+        assert y['Schema_version'] <= self.__class__.SCHEMA_VERSION
+        for item in y['items']:
+            if not keep_disabled and 'disable' in item and item['disable']:
+                continue
+            yield item
+
+    def parse(self):
+        for item in self._parse():
+            yield item[self.__class__.SCHEMA_PRIKEY]
+
+    def _write(self, callback):
+        d = {
+            'Schema': self.__class__.SCHEMA_VARIANT,
+            'Schema_version': self.__class__.SCHEMA_VERSION,
+            'items': sorted(
+                self._parse(keep_disabled=True),
+                key=lambda x: x[self.__class__.SCHEMA_PRIKEY])
+        }
+        callback(d)
+        with open(self._filename, 'w', encoding='utf-8') as f:
+            yaml.dump(d, f)
+
+    def _validate(self, item):
+        ip_regex = regex.compile(
+            r'(?(DEFINE)(?P<octet>'
+            '' r'1[0-9]{0,2}|2(?:[0-4][0-9])??|25[0-5]?|2[6-9]|[3-9][0-9]?))'
```

Comment:
I'm unsure what the purpose of the leading pair of single quotes here is...

Reply:
It's a hack to indent the line visually even though `flake8` boneheadedly wants to enforce its hamfisted convention. Maybe I'll have to add a comment.

Reply:
Perhaps [`(?x)`](https://docs.python.org/3/library/re.html#re.X) is a better solution than this hack.

Reply:
```python
        ip_regex = regex.compile(r'''
            (?(DEFINE)(?P<octet>
              1[0-9]{0,2}|2(?:[0-4][0-9])??|25[0-5]?|2[6-9]|[3-9][0-9]?))
            ^(?&octet)(?:\.(?&octet)){3}$''', regex.X
        )
```

Reply:
Good idea, thanks!